In [32]:
import scipy.stats as stats
import numpy as np
import math
import matplotlib.pyplot as plt

Задание 1 (на сопоставление)

In [ ]:
'''Ваш ответ на задание №1 (в формате 1-А, 2-В и т.д.) '''

1-E, 2-B, 3-F, 4-C, D, G, 5-A, G

Задание 2 (про порошок)

In [ ]:
'''Ваш ответ на задание №2 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- Будем использовать z-тест, т.к. известно стандартное отклонение генеральной совокупности неизвестно и мы имеем достаточное число наблюдений

- Гипотезы:
    - Нулевая гипотеза (H₀): Среднее значение порошка в коробке равно 3,00 кг (μ = 3,00)
    - Альтернативная гипотеза (H₁): Среднее значение порошка в коробке не равно 3,00 кг (μ ≠ 3,00)

In [4]:
# Данные из задачи

sample_mean = 3.005  # Средний вес образца
population_mean = 3.00  # Ожидаемое среднее значение
std_dev = 0.015  # Стандартное отклонение дозатора
n = 50  # Размер выборки
alpha = 0.05  # Уровень значимости


# Вычисление z-статистики
z_statistic = (sample_mean - population_mean) / (std_dev / n ** (0.5))

# Критическое значение для двухстороннего теста
z_critical = stats.norm.ppf(1 - alpha / 2)

# Вычисление p-value
p_value = 2 * (1 - stats.norm.cdf(abs(z_statistic)))

# Проверка гипотезы
if abs(z_statistic) > z_critical:
    result1 = "Нулевая гипотеза отклоняется. Дозатор требует настройки."
else:
    result1 = "Нет достаточных доказательств для отклонения нулевой гипотезы. Дозатор не требует настройки."

if p_value < alpha:
    result2 = "Нулевая гипотеза отклоняется. Дозатор требует настройки."
else:
    result2 = "Нет достаточных доказательств для отклонения нулевой гипотезы. Дозатор не требует настройки."

print(f"t-статистика: {z_statistic:.5f}")
print(f"\nКритическое значение: {z_critical:.5f}")
print(result1)
print(f"\np-value: {p_value:.5f}")
print(result2)

t-статистика: 2.35702

Критическое значение: 1.95996
Нулевая гипотеза отклоняется. Дозатор требует настройки.

p-value: 0.01842
Нулевая гипотеза отклоняется. Дозатор требует настройки.


Содержательгый вывод: на основе проведенного теста можно сделать вывод, что различия в дозировках есть и они статистически значимы. Таким образом, нулевая гипотеза о том, что среднее значение равно 3,00 кг, отклоняется, что свидетельствует о необходимости настройки дозатора.

Задание 3 (про двигатели)

In [ ]:
'''Ваш ответ на задание №3 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- Наиболее подходящим тестом будет двусторонний t-тест для двух независимых выборок, так как стандартные отклонения выборок известны, а генеральные — нет
- Гипотезы:
    - Нулевая гипотеза (H₀): Средние значения прочности веществ одинаковые (μ_1 = μ_2)
    - Альтернативная гипотеза (H₁): Средние значения прочности веществ различаются (μ_1 != μ_2)

In [12]:
n1, n2 = 15, 10
mean1, mean2 = 90, 88
std1, std2 = 2.3, 2.2
alpha = 0.05

# Расчёт t-статистики
t_calculated = (mean1 - mean2) / math.sqrt((std1**2 / n1) + (std2**2 / n2))

# Расчёт степеней свободы
df = ((std1**2 / n1 + std2**2 / n2) ** 2) / (
    ((std1**2 / n1) ** 2 / (n1 - 1)) + ((std2**2 / n2) ** 2 / (n2 - 1))
)

# Критическое значение t для двухстороннего теста
t_critical = stats.t.ppf(1 - alpha / 2, df)

# Вычисление p-value
p_value = 2 * (1 - stats.t.cdf(abs(t_calculated), df))

# Проверка гипотезы
print(f"Расчётное значение t-статистики: {t_calculated:.4f}\n")

print(f"Критическое значение t: {t_critical:.4f}")
if abs(t_calculated) > t_critical:
    result1 = "Отвергаем нулевую гипотезу, средние значения различаются."
else:
    result1 = "Не отвергаем нулевую гипотезу, средние значения не различаются."
print(result1)

print(f"\np-value: {p_value:.4f}")
if p_value < alpha:
    result2 = "Отвергаем нулевую гипотезу, средние значения различаются."
else:
    result2 = "Не отвергаем нулевую гипотезу, средние значения не различаются."
print(result2)

Расчётное значение t-статистики: 2.1865

Критическое значение t: 2.0856
Отвергаем нулевую гипотезу, средние значения различаются.

p-value: 0.0408
Отвергаем нулевую гипотезу, средние значения различаются.


Содержательный вывод: разницы между средними прочностными характеристиками вещеста не наблюдается

Задание 4 (про токи)

In [ ]:
'''Ваш ответ на задание №4 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- Для проверки гипотезы о равенстве средних значений используем двухвыборочный t-тест для независимых выборок. Мы предполагаем, что дисперсии равны, неизвестна дисперсия ген. совокупности.
- Гипотезы:
    - Нулевая гипотеза (H₀): Средний ток для обеих конструкций одинаков (μ_1 = μ_2)
    - Альтернативная гипотеза (H₁): Средние токи отличаются(μ_1 != μ_2)

In [8]:
tok_old = np.array([80.1, 82.3, 84.1, 82.6, 85.3, 81.3, 83.2, 81.7, 82.2, 81.4])
tok_new = np.array(
    [80.7, 81.3, 84.6, 81.7, 86.3, 84.3, 83.7, 84.7, 82.8, 84.4, 85.2, 84.9]
)

# Объемы выборок
n1 = len(tok_old)
n2 = len(tok_new)

# Средние значения и стандартные отклонения выборок
mean_old = np.mean(tok_old)
mean_new = np.mean(tok_new)
std_old = np.std(tok_old, ddof=1)
std_new = np.std(tok_new, ddof=1)

# Объединённая дисперсия (поскольку дисперсии равны)
s_p = np.sqrt((((n1 - 1) * std_old**2) + ((n2 - 1) * std_new**2)) / (n1 + n2 - 2))

# t-статистика для независимых выборок
t_stat_ind = (mean_old - mean_new) / (s_p * np.sqrt(1 / n1 + 1 / n2))

# Критическое значение t
t_critical_ind = stats.t.ppf(1 - 0.05 / 2, n1 + n2 - 2)

# p-value
p_value_ind = 2 * (1 - stats.t.cdf(abs(t_stat_ind), n1 + n2 - 2))


print(f"Среднее старой схемы {mean_old:.4f}\nСреднее новой схемы {mean_new:.4f}\n")
print(
    f"Стандартное отклонение старой схемы {std_old:.4f}\nСтандартное отклонение новой схемы {std_new:.4f}"
)
print(f'Расчётное значение t-статистики: {t_stat_ind:.4f}\n')

print(f"Критическое значение t: +-{t_critical_ind:.4f}")


if abs(t_stat_ind) > t_critical_ind:
    result1 = "Отвергаем нулевую гипотезу, средние значения различаются."
else:
    result1 = (
        "Не отвергаем нулевую гипотезу, нет значимой разницы между средними токами."
    )
print(result1)

print(f"\np-value: {p_value_ind:.4f}")
if p_value_ind < 0.05:
    result2 = "Отвергаем нулевую гипотезу, средние значения различаются."
else:
    result2 = (
        "Не отвергаем нулевую гипотезу, нет значимой разницы между средними токами."
    )
print(result2)

Среднее старой схемы 82.4200
Среднее новой схемы 83.7167

Стандартное отклонение старой схемы 1.4913
Стандартное отклонение новой схемы 1.7235
Расчётное значение t-статистики: -1.8657

Критическое значение t: +-2.0860
Не отвергаем нулевую гипотезу, нет значимой разницы между средними токами.

p-value: 0.0768
Не отвергаем нулевую гипотезу, нет значимой разницы между средними токами.


Содержательный вывод: производителю не следует заменять старую конструкцию схемы на новую

Задание 5 (про глюкозу)

In [ ]:
'''Ваш ответ на задание №5 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- так как мы анализируем малые выборки (10 наблюдений) и неизвестно стандартное отклонение ген. совокупности, то будет использовать парный t-тест
- Гипотезы:
    - Нулевая гипотеза (H₀): Нет разницы в уровнях глюкозы до и после приема препарата (μ_1 = μ_2)
    - Альтернативная гипотеза (H₁): Есть разница в уровнях глюкозы до и после приема препарата(μ_1 != μ_2)


In [9]:
x1 = np.array([9.6, 8.1, 8.8, 7.9, 9.2, 8.0, 8.4, 10.1, 7.8, 8.1])
x2 = np.array([7.7, 6.2, 7.4, 7.5, 8.3, 6.2, 8.1, 8.9, 7.4, 7.0])

# Разница между парами
differences = x1 - x2

# Выполнение парного t-теста
t_statistic, p_value = stats.ttest_1samp(differences, 0)

t_critical = stats.t.ppf(1 - alpha / 2, len(x1) - 1)

t_statistic, p_value, t_critical

print(f'Расчётное значение t-статистики: {t_statistic:.4f}\n')

print(f"Критическое значение t: +-{t_critical:.4f}")

if abs(t_statistic) > t_critical:
    result1 = "Отвергаем нулевую гипотезу"
else:
    result1 = (
        "Не отвергаем нулевую гипотезу"
    )
print(result1)

print(f"\np-value: {p_value:.4f}")
if p_value < 0.05:
    result2 = "Отвергаем нулевую гипотезу"
else:
    result2 = (
        "Не отвергаем нулевую гипотезу"
    )
print(result2)


Расчётное значение t-статистики: 5.7130

Критическое значение t: +-2.2622
Отвергаем нулевую гипотезу

p-value: 0.0003
Отвергаем нулевую гипотезу


Содержательный вывод: эффект от примема препарата на уровне доверия 0.95 есть

Задание 6 (про тросы)

In [ ]:
'''Ваш ответ на задание №6 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- Для проверки этих гипотез используем t-тест для двух независимых выборок с неравными дисперсиями (тест Уэлча), т.к. размер выборки небольшой и дисперсии выборок не равны по условию
- Гипотезы:
    - Нулевая гипотеза (H₀): Среднее напряжение разрушения нового кабеля равно среднему напряжению разрушения старого кабеля. (μ_new = μ_old)
    - Альтернативная гипотеза (H₁): Среднее напряжение разрушения нового кабеля больше, чем у старого.(μ_new > μ_old)


In [17]:
new_cable = np.array([92.7, 91.6, 94.7, 93.7, 96.5, 94.3, 93.7, 96.8, 98.9, 99.9])
old_cable = np.array([90.1, 92.2, 94.3, 92.1, 93.4, 91.1, 93.2, 91.5])

# Выполнение t-теста для двух независимых выборок с неравными дисперсиями (тест Уэлча)

t_statistic, p_value = stats.ttest_ind(new_cable, old_cable, equal_var=False)

# Вычисление степени свободы для теста Уэлча

n1, n2 = len(new_cable), len(old_cable)

var1, var2 = np.var(new_cable, ddof=1), np.var(old_cable, ddof=1)


df = ((var1 / n1 + var2 / n2) ** 2) / (
    ((var1 / n1) ** 2) / (n1 - 1) + ((var2 / n2) ** 2) / (n2 - 1)
)

# Рассчитываем критическое значение t для одностороннего теста на уровне значимости 1% (0.01)

alpha = 0.01

t_critical = stats.t.ppf(1 - alpha, df)

print(f"Расчётное значение t-статистики: {t_statistic:.4f}\n")

print(f"Критическое значение t: +-{t_critical:.4f}")
if abs(t_statistic) > t_critical:
    result1 = "Отвергаем нулевую гипотезу"
else:
    result1 = "Не отвергаем нулевую гипотезу"
print(result1)

print(f"\np-value: {p_value / 2:.4f}")
if p_value / 2 < alpha:
    result2 = "Отвергаем нулевую гипотезу"
else:
    result2 = "Не отвергаем нулевую гипотезу"
print(result2)

Расчётное значение t-статистики: 3.1204

Критическое значение t: +-2.6275
Отвергаем нулевую гипотезу

p-value: 0.0038
Отвергаем нулевую гипотезу


Содержательный вывод: изготовитель должен заменить старый кабель новым, так как мы отвергаем нулевую гипотезу

Задание 7 (про абитуриентов)

In [ ]:
'''Ваш ответ на задание №7 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- Используем парный t-test для связанных выборок, размер выборки меньше 30 и значение стандартного отклонения генеральной совокупности нам неизвестно
- Гипотезы:
  - Нулевая гипотеза (H₀): Подготовительные курсы не влияют на подготовку абитуриентов, то есть среднее изменение результата тестирования равно нулю.
  - Альтернативная гипотеза (H₁): Подготовительные курсы влияют на подготовку абитуриентов, то есть среднее изменение результата тестирования не равно нулю.

In [22]:
before = np.array([7, 6, 5, 4, 6, 2, 10, 3, 8, 5])
after = np.array([9, 6, 4, 5, 7, 4, 10, 6, 9, 6])

differences = after - before
alpha = 0.01

# Среднее и стандартное отклонение разностей
mean_diff = np.mean(differences)
std_diff = np.std(differences, ddof=1)
n = len(differences)

# t-статистика
t_statistic = mean_diff / (std_diff / np.sqrt(n))

# Критическое значение t для уровня значимости 0.01 (двусторонний тест)
alpha = 0.01
t_critical = stats.t.ppf(1 - alpha / 2, n - 1)

# p-value
p_value = 2 * (1 - stats.t.cdf(abs(t_statistic), n - 1))

print(f"Расчётное значение t-статистики: {t_statistic:.4f}\n")

print(f"Критическое значение t: +-{t_critical:.4f}")
if abs(t_statistic) > t_critical:
    result1 = "Отвергаем нулевую гипотезу"
else:
    result1 = "Не отвергаем нулевую гипотезу"
print(result1)

print(f"\np-value: {p_value:.4f}")
if p_value < alpha:
    result2 = "Отвергаем нулевую гипотезу"
else:
    result2 = "Не отвергаем нулевую гипотезу"
print(result2)

Расчётное значение t-статистики: 2.7386

Критическое значение t: +-3.2498
Не отвергаем нулевую гипотезу

p-value: 0.0229
Не отвергаем нулевую гипотезу


Содержательный вывод: подготовительные курсы на подготовку абитуриентов на уровне значимости 0.01 не влияют

Задание 8 (про фрукты и IQ)

In [ ]:
'''Ваш ответ на задание №8 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

In [27]:
oranges = np.array([125, 138, 112, 130, 124, 125, 110, 112, 117, 126])
pears = np.array([124, 102, 142, 130, 131, 124, 127, 115, 108, 109])

# Вычисление дисперсий
var_oranges = np.var(oranges, ddof=1)
var_pears = np.var(pears, ddof=1)

# Статистика F
F_statistic = var_oranges / var_pears if var_oranges > var_pears else var_pears / var_oranges

# Степени свободы
df_oranges = len(oranges) - 1
df_pears = len(pears) - 1

# Критическое значение F для уровня значимости 0.05
alpha = 0.05
F_critical = stats.f.ppf(1 - alpha, df_oranges, df_pears)

# p-value
p_value = 2 * (1 - stats.f.cdf(F_statistic, df_oranges, df_pears))

F_statistic, F_critical, p_value

print(f"Расчётное значение F-статистики: {F_statistic:.4f}\n")

print(f"Критическое значение F: {F_critical:.4f}")
if abs(F_statistic) > F_critical:
    result1 = "Отвергаем нулевую гипотезу"
else:
    result1 = "Не отвергаем нулевую гипотезу"
print(result1)

print(f"\np-value: {p_value / 2:.4f}")
if p_value / 2 < alpha:
    result2 = "Отвергаем нулевую гипотезу"
else:
    result2 = "Не отвергаем нулевую гипотезу"
print(result2)


Расчётное значение F-статистики: 1.9062

Критическое значение F: 3.1789
Не отвергаем нулевую гипотезу

p-value: 0.1753
Не отвергаем нулевую гипотезу


Содержательный вывод: дисперсии обеих групп не одинаковые на уровне значимости 0.05

Задание 9 (про самолеты)

In [ ]:
'''Ваш ответ на задание №9 
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

Можно предположить, что данные имеют геометрическое распределение:
- На каждом вылете вероятность найти корабль (успех) примерно одинаковая, если не учитывать внешние факторы
- Мы фиксируем количество вылетов, которые потребовались для нахождения корабля, и считаем количество попыток до первого успеха
- Каждый вылет самолета — это независимое событие
- Гипотезы:
  - Нулевая гипотеза: данные распределены геометрически
  - Альтернативная гипотеза: данные не распределены геометрически

In [37]:
X = np.array([1, 2, 3, 4, 5, 6, 7, 8, 9])  # Количество вылетов
Y = np.array([36, 23, 11, 13, 8, 3, 2, 1, 3])  # Количество успешных попыток

alpha = 0.05

# Вычисление суммы наблюдений
n = Y.sum()

# Оценка параметра p для геометрического распределения
p_estimated = n / np.dot(X, Y)
print(f"Оценочная вероятность p = {p_estimated}")

# Расчет ожидаемых частот (Y_p) на основе геометрического распределения
Y_p = stats.geom.pmf(X, p_estimated) * n

# орректировка на возможное несоответствие суммы наблюдаемых и ожидаемых частот
correction = (n - Y_p.sum()) / len(Y)
Y_p += correction  # Корректируем расчетные значения

# Проведение хи-квадрат теста для сравнения наблюдаемых и ожидаемых частот
chi2_stat, p_value = stats.chisquare(Y, Y_p, ddof=1)

# Степени свободы для теста (df = количество категорий - 1 - 1, где 1 — параметр p)
df = Y.shape[0] - 1 - 1

# Критическое значение для хи-квадрат распределения
chi2_critical = stats.chi2.ppf(1 - alpha, df)

print(f"Хи-квадрат статистика: {chi2_stat:.4f}")

print(f"\nКритическое значение хи-квадрат: {chi2_critical:.4f}")

if chi2_stat > chi2_critical:
    print("Отклонить нулевую гипотезу на основе хи-квадрат статистики")
else:
    print("Нельзя отклонить нулевую гипотезу на основе хи-квадрат статистики")

print(f"\np-value: {p_value:.4f}")
if p_value < alpha:
    print("Отклонить нулевую гипотезу на основе p-значения")
else:
    print("Нельзя отклонить нулевую гипотезу на основе p-значения")

Оценочная вероятность p = 0.36496350364963503
Хи-квадрат статистика: 6.4975

Критическое значение хи-квадрат: 14.0671
Нельзя отклонить нулевую гипотезу на основе хи-квадрат статистики

p-value: 0.4830
Нельзя отклонить нулевую гипотезу на основе p-значения


Содержательный вывод: нет оснований считать, что данные распределены не геометрически

Задание 10 (про цвет глаз и волос)

In [ ]:
'''Ваш ответ на задание №10
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено критическое значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения критического значения с расчетным
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- Для проверки гипотезы о связи между цветом глаз и цветом волос используем критерий Хи-квадрат для независимости, он позволяет проверить, есть ли статистически значимая зависимость между двумя категориальными переменными
- Гипотезы:
    - Нулевая гипотеза (H₀): Цвет глаз и цвет волос независимы.
    - Альтернативная гипотеза (H₁): Цвет глаз и цвет волос зависимы.

In [39]:
observed = np.array([
    [35, 52, 13],
    [28, 42, 10],
    [7, 10, 3]
])

# Ожидаемые частоты
row_sums = np.sum(observed, axis=1)
col_sums = np.sum(observed, axis=0)
total_sum = np.sum(observed)

expected = np.outer(row_sums, col_sums) / total_sum

# Хи-квадрат статистика
chi2_statistic = np.sum((observed - expected) ** 2 / expected)

# Степени свободы (число строк - 1) * (число столбцов - 1)
df = (observed.shape[0] - 1) * (observed.shape[1] - 1)

# Критическое значение хи-квадрат для уровня значимости 0.05
alpha = 0.05
chi2_critical = stats.chi2.ppf(1 - alpha, df)

# p-value
p_value = 1 - stats.chi2.cdf(chi2_statistic, df)

chi2_statistic, chi2_critical, p_value

print(f"Расчётное значение Хи-квадрат-статистики: {chi2_statistic:.4f}\n")

print(f"Критическое значение Хи-квадрат: {chi2_critical:.4f}")
if abs(chi2_statistic) > chi2_critical:
    result1 = "Отвергаем нулевую гипотезу"
else:
    result1 = "Не отвергаем нулевую гипотезу"
print(result1)

print(f"\np-value: {p_value :.4f}")
if p_value  < alpha:
    result2 = "Отвергаем нулевую гипотезу"
else:
    result2 = "Не отвергаем нулевую гипотезу"
print(result2)


Расчётное значение Хи-квадрат-статистики: 0.0962

Критическое значение Хи-квадрат: 9.4877
Не отвергаем нулевую гипотезу

p-value: 0.9989
Не отвергаем нулевую гипотезу


Содержательный вывод: статистически значимой связи между цветом глаз и волос не обнаружено

Задание 11 (про тигров)

In [ ]:
'''Ваш ответ на задание №11
- верно указана применяемая статистика и обоснование ее применения
- получено расчетное значение статистики
- получено значение p_value
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- в случае, если нулевая гипотеза отвергнута, сделан тест Тьюки для подробного сравнения средних
- сделано дополнение к статистическому выводу по результатам теста Тьюки
- сделан содержательный вывод по поставленной проблеме'''

- Для оценки влияния типа корма на среднюю массу тигров в данном случае применим анализ дисперсии (ANOVA), тест позволяет определить, есть ли статистически значимые различия в средних значениях массы тигров, кормящихся разными типами корма
- Гипотезы:
  - Нулевая гипотеза (H₀): Средние массы тигров, кормящихся разными типами корма, одинаковы.
  - Альтернативная гипотеза (H₁): Средние массы тигров, кормящихся разными типами корма, различаются.

In [46]:
from statsmodels.stats.multicomp import pairwise_tukeyhsd

In [47]:
X1 = np.array([151, 135, 137, 118, 132, 135, 131, 121, 137, 140, 152, 133, 151, 132])
X2 = np.array([108, 123, 145, 103, 110, 124, 116, 108, 104, 130, 106, 105, 114, 108])
X3 = np.array([140, 152, 149, 132, 124, 146, 156, 137, 124, 145, 147, 139, 157, 139])
X4 = np.array([127, 135, 110, 116, 119, 123, 102, 121, 118, 120, 114, 134, 121, 109])

# Применение однофакторного ANOVA
f_statistic, p_value_anova = stats.f_oneway(X1, X2, X3, X4)

# Если p-value < 0.05, проводим тест Тьюки
if p_value_anova < 0.05:

    data = np.concatenate([X1, X2, X3, X4])
    labels = ["X1"] * len(X1) + ["X2"] * len(X2) + ["X3"] * len(X3) + ["X4"] * len(X4)
    tukey_result = pairwise_tukeyhsd(data, labels, alpha=0.05)
    print("\nРезультаты теста Тьюки:", tukey_result)
    print(f"F-статистика: {f_statistic:.4f}")
    print(f"P-value: {p_value_anova:}")
else:
    tukey_result = None


Результаты теста Тьюки:  Multiple Comparison of Means - Tukey HSD, FWER=0.05  
group1 group2 meandiff p-adj   lower    upper   reject
------------------------------------------------------
    X1     X2    -21.5    0.0  -32.019  -10.981   True
    X1     X3   5.8571  0.458  -4.6619  16.3762  False
    X1     X4 -16.8571 0.0005 -27.3762  -6.3381   True
    X2     X3  27.3571    0.0  16.8381  37.8762   True
    X2     X4   4.6429 0.6473  -5.8762  15.1619  False
    X3     X4 -22.7143    0.0 -33.2333 -12.1953   True
------------------------------------------------------
F-статистика: 21.9278
P-value: 2.5927802386467297e-09


На имеющихся данных, на уровне значимости 5% есть достаточные основания отвергнуть нулевую гипотезу в пользу альтернативной

X1 и X2: Значимая разница (p-value = 0.0), что означает, что масса тигров, кормящихся кормом X1, значительно отличается от массы тигров, кормящихся кормом X2  
X1 и X3: Разница незначима (p-value = 0.458), т.е. массы тигров, кормящихся кормами X1 и X3, не имеют статистически значимого отличия  
X1 и X4: Значимая разница (p-value = 0.0005), то есть масса тигров, кормящихся кормом X1, значительно отличается от массы тигров, кормящихся кормом X4  
X2 и X3: Значимая разница (p-value = 0.0), то есть масса тигров, кормящихся кормом X2, значительно отличается от массы тигров, кормящихся кормом X3  
X2 и X4: Разница незначима (p-value = 0.6473), массы тигров, кормящихся кормами X2 и X4, не имеют статистически значимого отличия  
X3 и X4: Значимая разница (p-value = 0.0), масса тигров, кормящихся кормом X3, значительно отличается от массы тигров, кормящихся кормом X4  

Задание 12 (про белок)

In [ ]:
'''Ваш ответ на задание №12
- верно указана применяемая статистика и обоснование ее применения
- расчет выполнен через последовательные суммы квадратов
- получены расчетные значения статистики
- получены значения p_value
- сделан статистический вывод на основе сравнения p-value и уровня значимости
- сделан содержательный вывод по поставленной проблеме'''

- Будем использовать анализ дисперсии (ANOVA) с применением метода последовательных сумм квадратов,он позволяет анализировать влияние нескольких факторов на зависимую переменную, например, на массу тела в данном случае
- Гипотезы:
  - Нулевая гипотеза (H₀): Возраст и наличие потомства (размножение) не влияют на массу тела белок.
  - Альтернативная гипотеза (H₁): Возраст и наличие потомства влияют на массу тела белок.


In [50]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
import pandas as pd

In [ ]:
data = {
    'has_brood': [0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1],
    'age': [1,1,1,1,1,2,2,2,2,2,3,3,3,3,3,1,1,1,1,1,2,2,2,2,2,3,3,3,3,3],
    'weight': [440,438,429,502,602,892,868,855,866,932,1575,849,759,1602,1327,308,328,326,326,325,737,798,876,810,861,1001,901,958,1032,883]
}

df = pd.DataFrame(data)

# Модель ANOVA с факторами age и has_brood, и их взаимодействием
model = ols('weight ~ C(age) + C(has_brood) + C(age):C(has_brood)', data=df).fit()

# Выполнение анализа дисперсии
anova_table = sm.stats.anova_lm(model, typ=2)

anova_table

sum_sq    df          F        PR(>F)
C(age)               2427075.8   2.0  42.377707  1.333903e-08
C(has_brood)          202705.2   1.0   7.078626  1.368668e-02
C(age):C(has_brood)    50688.2   2.0   0.885036  4.257428e-01
Residual              687269.6  24.0        NaN           NaN

Мы отвергаем нулевую гипотезу о том, что возраст и наличие потомства не влияют на массу тела на уровне значимости 5%

Содержательный вывод:  Возраст и наличие потомства оказывают влияние на массу тела самок африканских земляных белок. Однако, их взаимодействие не является значимым фактором для массы тела, что указывает на то, что эти два фактора действуют независимо друг от друга.